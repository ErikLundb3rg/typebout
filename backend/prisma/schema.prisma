// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Users {
  id           Int            @id @default(autoincrement())
  name         String
  password     String
  createdAt    DateTime       @default(now())
  performances Performances[]
}

model Performances {
  id           Int      @id @default(autoincrement())
  correct      Int
  mistakes     Int
  completed_in Int
  createdAt    DateTime @default(now())
  user         Users    @relation(fields: [userId], references: [id])
  userId       Int
  race         Races    @relation(fields: [raceId], references: [id])
  raceId       Int
}

model Races {
  id           Int            @id @default(autoincrement())
  createdAt    DateTime       @default(now())
  performances Performances[]
  quote        Quotes         @relation(fields: [quotesId], references: [id])
  quotesId     Int
}

model Quotes {
  id         Int          @id @default(autoincrement())
  content    String
  categories Categories[]
  author     Author       @relation(fields: [authorId], references: [id])
  authorId   Int
  Races      Races[]
}

model Author {
  id     Int      @id @default(autoincrement())
  name   String
  Quotes Quotes[]
}

model Categories {
  id     Int      @id @default(autoincrement())
  name   String
  quotes Quotes[]
}
