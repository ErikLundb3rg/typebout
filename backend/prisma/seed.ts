import { PrismaClient } from '@prisma/client'
import quoteFile from '../quotes/quotes.json'

const prisma = new PrismaClient()

const main = async () => {
  // delete all quotes

  quoteFile.quotes.forEach(async (quote) => {
    await prisma.quotes.upsert({
      where: { content: quote.content },
      update: {},
      create: {
        content: quote.content,
        author: {
          connectOrCreate: {
            where: {
              name: quote.author
            },
            create: {
              name: quote.author
            }
          }
        }
      }
    })
  })
}

main()
  .then(async () => {
    await prisma.$disconnect()
  })
  .catch(async (error) => {
    console.error(error)
    await prisma.$disconnect()
    process.exit(1)
  })
