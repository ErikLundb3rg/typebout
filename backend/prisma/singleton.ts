import { PrismaClient } from '@prisma/client';
import { mockDeep, mockReset, DeepMockProxy } from 'jest-mock-extended';

import db from './client';

jest.mock('./client', () => ({
  __esModule: true,

  default: mockDeep<PrismaClient>()
}));

export const prismaMock = db as unknown as DeepMockProxy<PrismaClient>;

beforeEach(() => {
  mockReset(prismaMock);
});
