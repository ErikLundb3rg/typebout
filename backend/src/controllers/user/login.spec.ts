/* eslint-disable @typescript-eslint/no-explicit-any */
import { prismaMock } from '../../../prisma/singleton'
import bcrypt from 'bcrypt'
import { login } from './index'
import {
  generateAccessToken,
  generateRefreshToken
} from '../../auth/util/verifyers'

describe('Login Controller', () => {
  let req: any = {}
  let res: any = {}
  let user: any = {}

  beforeEach(() => {
    req = {}
    res = {}
    user = {
      id: 1,
      username: 'username',
      password: 'password',
      createdAt: new Date()
    }
    res.cookie = jest.fn()
    ;(generateAccessToken as jest.Mock) = jest
      .fn()
      .mockReturnValue('accesstoken')
    ;(generateRefreshToken as jest.Mock) = jest.fn()
    prismaMock.users.findUnique.mockResolvedValue(user)
  })

  test('is successful with correct input', async () => {
    req.body = {
      username: 'username',
      password: 'password'
    }
    ;(bcrypt.compare as jest.Mock) = jest.fn().mockResolvedValue(true)

    expect(await login(req, res)).toMatchSnapshot({
      data: {
        user: {
          createdAt: expect.any(Date)
        }
      }
    })
  })

  test('is unsuccessful with faulty password', async () => {
    req.body = {
      username: 'username',
      password: 'wrong_password'
    }
    ;(bcrypt.compare as jest.Mock) = jest.fn().mockResolvedValue(false)

    expect(await login(req, res)).toMatchSnapshot()
  })

  test('is unsuccessful with non-existant username', async () => {
    req.body = {
      username: 'wrong_username',
      password: 'password'
    }
    ;(bcrypt.compare as jest.Mock) = jest.fn()

    expect(bcrypt.compare).toHaveBeenCalledTimes(0)

    expect(await login(req, res)).toMatchSnapshot()
  })
})
