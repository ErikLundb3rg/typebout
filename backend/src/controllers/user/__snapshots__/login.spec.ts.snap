// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Controller is successful with correct input 1`] = `
{
  "data": {
    "accessToken": "accesstoken",
  },
  "message": "Successful Login",
  "ok": true,
  "status": 200,
}
`;

exports[`Login Controller is unsuccessful with faulty password 1`] = `
{
  "data": null,
  "message": "Could not find a user with that username or password",
  "ok": false,
  "status": 400,
}
`;

exports[`Login Controller is unsuccessful with non-existant username 1`] = `
{
  "data": null,
  "message": "Could not find a user with that username or password",
  "ok": false,
  "status": 400,
}
`;
